### PR TITLE
cmake/AddCephTest: bind crimson unittest to different cores

### PR DESCRIPTION
**Pull request by rosinL**
When running crimson unittest, the seastar framework always
use and only use cpu0, and with many parallel crimson unittest
jobs, all the jobs are running on cpu0, the other cpu cores
can't used, make the make check run very slow, even timeout
happens. Use set_property RESOURCE_GROUPS to specify cpu resources
to crimson unittest, and accelerate make check running.

Fixes: https://tracker.ceph.com/issues/64117

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by rosinL** [comment]
jenkins test make check arm64

**Comment by rosinL** [comment]
For x86 make check， the longest crimson unittest job `unittest-transaction-manager`, costtime reduced from 1940.81 sec to 983.14 sec. As long tail job `run-rbd-unit-tests-61.sh`, the total test time reduced not so much. As make check arm is still running ,there is no Arm test data for now, but it does same effect on Arm.
Before(x86)
```
287/290 Test #257: unittest-omap-manager .....................   Passed  1642.30 sec
288/290 Test  #32: run-rbd-unit-tests-61.sh ..................   Passed  1728.38 sec
289/290 Test #264: unittest-fltree-onode-manager .............   Passed  1688.28 sec
290/290 Test #251: unittest-transaction-manager ..............   Passed  1940.81 sec

100% tests passed, 0 tests failed out of 290

Total Test time (real) = 1980.95 sec
``` 
After(x86)
```
287/290 Test #251: unittest-transaction-manager ..............   Passed  983.14 sec
288/290 Test  #31: run-rbd-unit-tests-1.sh ...................   Passed  1529.75 sec
289/290 Test #147: check-generated.sh ........................   Passed  1686.12 sec
290/290 Test  #32: run-rbd-unit-tests-61.sh ..................   Passed  1689.13 sec

100% tests passed, 0 tests failed out of 290

Total Test time (real) = 1689.17 sec
```

**Comment by rosinL** [comment]
@ljflores @athanatos @Matan-B @rzarzynski @cyx1231st Please help to review, thanks.

**Comment by athanatos** [comment]
@rosinL So different crimson unittest processes are being bound to core 0?  I'd have expected --smp 0 to limit any particular process to a single reactor, but not to limit it to any specific core.

**Comment by rosinL** [comment]
jenkins test make check arm64

**Comment by rosinL** [comment]
@athanatos As I test, with --smp 0 , the test will report `runtime error: division by zero`. Are there some mistakes, please let me know.

**Comment by rosinL** [comment]
I run 2 crimson unittest job parallelly
Without `--cpuset`, all the jobs run on cpu0
```
root@ubuntu:/home# ps -ef |grep unittest
root     3373367 1082838 59 01:27 pts/1    00:00:02 ./build/bin/unittest-fltree-onode-manager --memory 256M --smp 1
root     3373370  251881 56 01:27 pts/2    00:00:02 ./bin/unittest-transaction-manager --memory 256M --smp 1
root     3373374  715758  0 01:27 pts/4    00:00:00 grep --color=auto unittest
root@ubuntu:/home# taskset -acp 3373367
pid 3373367's current affinity list: 0-47
pid 3373368's current affinity list: 0
pid 3373369's current affinity list: 0
root@ubuntu:/home# taskset -acp 3373370
pid 3373370's current affinity list: 0-47
pid 3373371's current affinity list: 0
pid 3373372's current affinity list: 0
``` 
With `--cpuset` to separate jobs to different cores
```
root@ubuntu:/home# ps -ef |grep unittest
root     3373379 1082838 93 01:28 pts/1    00:00:10 ./build/bin/unittest-fltree-onode-manager --memory 256M --smp 1 --cpuset 1
root     3373382  251881 96 01:28 pts/2    00:00:06 ./bin/unittest-transaction-manager --memory 256M --smp 1 --cpuset 0
root     3373386  715758  0 01:28 pts/4    00:00:00 grep --color=auto unittest
root@ubuntu:/home# taskset -acp 3373379
pid 3373379's current affinity list: 0-47
pid 3373380's current affinity list: 1
pid 3373381's current affinity list: 1
root@ubuntu:/home# taskset -acp 3373382
pid 3373382's current affinity list: 0-47
pid 3373383's current affinity list: 0
pid 3373384's current affinity list: 0
```

**Comment by rosinL** [comment]
@athanatos If there are other better solutions, please let me know, thanks.

**Comment by rosinL** [comment]
jenkins test make check arm64

**Review by tchaikov — COMMENTED** [review]
@rosinL Hi Rixin, thank you for your efforts to improve the efficiency of crimson tests. i'd suggest taking a look at https://cmake.org/cmake/help/latest/manual/ctest.1.html#ctest-resource-allocation, and consider CPU shards as a resource. actually, the same applies to the memory.

**Comment by rosinL** [comment]
@tchaikov I will do some research and have a try

**Comment by rosinL** [comment]
jenkins test make check arm64

File: cmake/modules/AddCephTest.cmake [hunk 1]
```diff
@@ -2,7 +2,36 @@
 
 #adds makes target/script into a test, test to check target, sets necessary environment variables
 function(add_ceph_test test_name test_path)
-  add_test(NAME ${test_name} COMMAND ${test_path} ${ARGN}
+  # crimson seastar unittest always run with --memory XXXM --smp N
+  string(FIND "${ARGN}" "--memory" mem_pos)
+  string(FIND "${ARGN}" "--smp" smp_pos)
+  if((NOT mem_pos EQUAL -1) AND (NOT smp_pos EQUAL -1))
```
Comment (cyx1231st) [inline cmake/modules/AddCephTest.cmake:8]: I think it is probably not a good assumption that all crimson tests always run with "--memory", and this feature doesn't seem to be related to the memory configuration. Is it fine to drop the memory related checks?

Otherwise, the cpu allocation policy below seems good to me.

Comment (rosinL) [reply cmake/modules/AddCephTest.cmake:8]: Ok, I will remove the `--memory` checks

**Review by cyx1231st — APPROVED** [review]
The --cpuset configuration policy looks good to me to distribute the crimson tests to all the available cores in a round-robin way.

**Comment by rosinL** [comment]
@tchaikov According my understand,  we need `taskset or numactl or seastar arg --cpuset` to distribute crimson tests to different cores. ctest resource allocation can consider CPU shards as a resource, but after CPU shard allocated, ctest cann't bind crimson test to the CPU shard automatically, still need `taskset or numactl or seastar arg --cpuset` to bind cores manually, It seems more  complicated, Am I right?

**Comment by rosinL** [comment]
jenkins test api

**Comment by rosinL** [comment]
As make check(arm64) running pretty slow, Can we merge this or move a step forward if needed? @ljflores

File: cmake/modules/AddCephTest.cmake [hunk 2]
```diff
@@ -2,7 +2,35 @@
 
 #adds makes target/script into a test, test to check target, sets necessary environment variables
 function(add_ceph_test test_name test_path)
-  add_test(NAME ${test_name} COMMAND ${test_path} ${ARGN}
+  # crimson seastar unittest always run with --smp N
+  string(FIND "${ARGN}" "--smp" smp_pos)
```
Comment (Matan-B) [inline cmake/modules/AddCephTest.cmake:6]: seastar::smp::count is determined based on `--cpuset` size provided. We can avoid setting both of them.

Comment (rosinL) [reply cmake/modules/AddCephTest.cmake:6]: > seastar::smp::count is determined based on --cpuset size provided 

Yes,  but when we use `add_ceph_unittest(unittest-seastar-config --memory 256M --smp 4)` to add crimson test, we just know how many cores needed, it is complicated to decide `--cpuset`, so we add `--cpuset` automatically according `--smp`,  it is unnecessary to remove `--smp`  from the exists args

Comment (Matan-B) [reply cmake/modules/AddCephTest.cmake:6]: seastar's `cpuset` defaults to "all" cores. Can we set only `--smp` in our tests?

Comment (rosinL) [reply cmake/modules/AddCephTest.cmake:6]: No, without `cpuset`, it will start from cpu0, if there are many jobs, cpu0 will be shared and running pretty slow.

**Comment by rosinL** [comment]
> ctest populates the allocated resource using environment variables like CTEST_RESOURCE_GROUP_<num>_<resource-type>. i'd suggest rework the crimson tests to respect these variables, and set the seastar application config accordingly

@tchaikov Do you mean like this:
```
int main(int argc, char** argv) {
    seastar::app_template app;
    // get cpuset from CTEST_RESOURCE_GROUP_<num>_<resource-type>
    // get new_argv with argc + 2  elements
    // copy argv to new_argv
    // new_argv[argc]="--cpuset"
    // new_argv[argc+1]="startcpuid-endcpuid"
    app.run(argc, new_argv, [] {
        return crimson_test_func();
    });
}
```

**Comment by tchaikov** [comment]
> > ctest populates the allocated resource using environment variables like CTEST_RESOURCE_GROUP__. i'd suggest rework the crimson tests to respect these variables, and set the seastar application config accordingly
> 
> @tchaikov Do you mean like this:
> 
> ```
> int main(int argc, char** argv) {
>     seastar::app_template app;
>     // get cpuset from CTEST_RESOURCE_GROUP_<num>_<resource-type>
>     // get new_argv with argc + 2  elements
>     // copy argv to new_argv
>     // new_argv[argc]="--cpuset"
>     // new_argv[argc+1]="startcpuid-endcpuid"
>     app.run(argc+2, new_argv, [] {
>         return crimson_test_func();
>     });
> }
> ```

something like:

```c++
struct ctest_resource {
  int id;
  int slots;
};

std::vector<ctest_resource> parse_ctest_resources(const std::string resource_spec) {
  vector<string> resources;
  boost::split(resources, resource_spec, line,boost::is_any_of(";"));
  std::regex res_regex("id:([0-9]+),slots:([0-9]+)");
  vector<ctest_resource> ctest_resources;
  for (auto& resource : resources) {
    std::smatch matched;
    std::regex_match(resource, matched, res_regex));
    int id = stoi(matched[0].str());
    int slots = stoi(matched[1].str());
    ctest_resources.emplace_back(id, slots);
  }
  return ctest_resources;
}

std::optional<seastar::resource::cpuset> cpuset_from_ctest_resource_group() {
  int nr_groups = 0;
  if (auto group_count = getenv("CTEST_RESOURCE_GROUP_COUNT"); group_count != nullptr) {
    nr_groups = std::stoi(group_count);
  } else {
    return {};
  }
  seastar::resource::cpuset cpuset;
  for (int num = 0; num < nr_groups; num++) {
    std::string resource_type_name;
    fmt::format_to(std::back_inserter(resource_type_name), "CTEST_RESOURCE_GROUP_{}", num);
    // only a single resource type is supported for now
    std::string resource_type = getenv(resource_type_name.data());
    std::transform(resource_type.begin(), resource_type.end(), resource_type.begin(), ::toupper);
    std::string resource_group;
    fmt::format_to(std::back_inserter(resource_group), "CTEST_RESOURCE_GROUP_{}_{}", num, resource_type);
    std::string resource_spec = getenv(resource_group.data());
    if (group == "CPUS") {
      for (auto& resource : parse_ctest_resources(resource_spec)) {
        // each id has a single cpu slot
        cpuset.insert(resource.id);
      }
    } else {
      fmt::print(cerr, "unsupported resource type: {}", resource_type);
    }
  }
  return cpuset;
}

int main(int argc, char** argv) {
  app_template::seastar_options opts;
  if (auto cpuset = cpuset_from_ctest_resource_group(); cpuset)
    opts.smp_opts.cpuset.set_value(*cpuset);
  }
  app_template app{std::move(opts)};
  return app.run(argc, argv, [] {
    return run_tests();
  };
}
```

**Comment by rosinL** [comment]
jenkins test make check

File: cmake/modules/AddCephTest.cmake [hunk 3]
```diff
@@ -22,6 +22,22 @@ function(add_ceph_test test_name test_path)
...
+      math(EXPR i "${i} + 1")
+      if(argv_value STREQUAL "--smp")
+        list(GET ARGV ${i} smp_count)
+        break()
+      endif()
+    endwhile()
+    set(resource_groups "${smp_count},cpus:1")
+    set_property(TEST ${test_name}
+      PROPERTY RESOURCE_GROUPS ${resource_groups})
+  endif()
```
Comment (tchaikov) [inline cmake/modules/AddCephTest.cmake:40]: this is not quite elegant. we could improve it in two ways.  now that we already know the index of `--smp` in the `${ARGN}`. why not just get its option with something like:
```cmake
if(smp_pos GREATER -1)
  if(smp_pos EQUAL ARGC) 
    message(FATAL_ERROR "${test_name} --smp requires an argument")
  endif()
  math(EXPR i "${i} + 1")
  list(GET ARGV ${i} smp_count)
  # we assign CPU cores as resources to each tests with "--smp" option.
  # each group represents a core of CPU, and each group has only a single
  # slot. here, the test in question requires:
  # ${smp_count} groups, each of which requires 1 slot.
  set_property(TEST ${test_name}
       PROPERTY RESOURCE_GROUPS "${smp_count},cpus:1")
endif()
```
or we can use  update the tests to pass smp and memory with keyword argument, and use `cmake_parse_arguments(). that'd be a more intrusive change.

so i'd suggest go with the first approach. the point is to add more comment to explain the concept of group and slot. and how they maps to the `--smp` option.

File: run-make-check.sh [hunk 4]
```diff
@@ -42,15 +62,18 @@ function run() {
         $DRY_RUN sudo /sbin/sysctl -q -w fs.aio-max-nr="${aiomax}"
     fi
 
+    CTEST_RESOURCE_FILE=`mktemp /tmp/ctest-resource-XXXXXX`
```
Comment (tchaikov) [inline run-make-check.sh:65]: could let `gen_ctest_resource_file()` `mktemp ...` and `echo` the filename. simpler this way, something like:

```sh
function gen_ctest_resource_file () {
  local filename=`mktemp /tmp/ctest-resource-XXXXXX`
  # write to $filename
  echo "$filename"
}

# ...
CHECK_MAKEOPTS+=" --resource-spec-file $(gen_ctest_resource_file)"
```

File: run-make-check.sh [hunk 5]
```diff
@@ -22,6 +22,26 @@ source src/script/run-make.sh
...
+    do
+        if [ $i -eq ${max_cpuid} ]; then
+            echo -e "        {\n          \"id\": \"$i\",\n          \"slots\": 1\n        }" >> ${CTEST_RESOURCE_FILE}
+        else
+            echo -e "        {\n          \"id\": \"$i\",\n          \"slots\": 1\n        }," >> ${CTEST_RESOURCE_FILE}
+        fi
+    done
+    echo -e "${cpu_group_end}" >> ${CTEST_RESOURCE_FILE}
+    echo -e "${json_end}" >> ${CTEST_RESOURCE_FILE}
+}
```
Comment (tchaikov) [inline run-make-check.sh:43]: this function can be rewritten using jq, like
```sh
local max_cpuid=$(($(nproc) - 1))
jq -n '$ARGS.positional | map({"id":.,slots:1}) | {version: {major:1,minor:0},local: .}' --args $(seq $max_cpuid) > $filename
```

File: src/test/crimson/ctest_seastar.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```
Comment (tchaikov) [inline src/test/crimson/ctest_seastar.h:1]: the major part of this file was created by me. could please use my Signed-off-by ?

File: src/test/crimson/ctest_seastar.h [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  return cpuset;
+}
+
+static seastar::app_template::seastar_options get_smp_opts_from_ctest() {
+  seastar::app_template::seastar_options opts;
+  auto cpuset = get_cpuset_from_ctest_resource_group();
+  if (cpuset)
+    opts.smp_opts.cpuset.set_value(*cpuset);
+  return opts;
+}
```
Comment (tchaikov) [inline src/test/crimson/ctest_seastar.h:76]: and please squash this commit into 1e2195a152d6aa45ebf18d6a896aff8ec65d0b0c, so it's more clear how this file is used.

File: src/test/crimson/ctest_seastar.h [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <vector>
+#include <string>
+#include <regex>
+#include <optional>
+#include <cstdlib>
+#include <iostream>
```
Comment (tchaikov) [inline src/test/crimson/ctest_seastar.h:11]: could you please order the include in alphabetical order?

File: src/test/crimson/ctest_seastar.h (hunk 7 above)
Comment (tchaikov) [reply src/test/crimson/ctest_seastar.h:76]: also, i'd suggest rename this header file to something like `ctest_utils.h` or `ctest_resource.h`.

**Review by tchaikov — COMMENTED** [review]
left couple comments. also, please explain the motive in the related commit message. and add paste the test results before and after the change to show the improvement contributed by this change.

**Comment by rosinL** [comment]
jenkins test make check

**Comment by rosinL** [comment]
jenkins retest this please

**Comment by rosinL** [comment]
For x86 irvingi07 with 24 cores, the Total Test time keep the same as the long tail job `check-generated.sh`,  the longest crimson unittest `unittest-transaction-manager` job's time decreased from 2522.75 sec to 1375.03 sec, nearly 2X performance improvement.
Before
```
285/290 Test #257: unittest-omap-manager .....................   Passed  2010.54 sec
286/290 Test  #32: run-rbd-unit-tests-61.sh ..................   Passed  2425.26 sec
287/290 Test #148: readable.sh ...............................   Passed  2458.99 sec
288/290 Test #264: unittest-fltree-onode-manager .............   Passed  2093.77 sec
289/290 Test #251: unittest-transaction-manager ..............   Passed  2522.75 sec
290/290 Test #147: check-generated.sh ........................   Passed  2846.91 sec

100% tests passed, 0 tests failed out of 290

Total Test time (real) = 2919.83 sec
```
After
```
285/290 Test  #52: unittest_bufferlist .......................   Passed  1238.60 sec
286/290 Test #251: unittest-transaction-manager ..............   Passed  1375.03 sec
287/290 Test  #31: run-rbd-unit-tests-1.sh ...................   Passed  2063.50 sec
288/290 Test  #32: run-rbd-unit-tests-61.sh ..................   Passed  2481.31 sec
289/290 Test #148: readable.sh ...............................   Passed  2446.87 sec
290/290 Test #147: check-generated.sh ........................   Passed  2838.97 sec

100% tests passed, 0 tests failed out of 290

Total Test time (real) = 2914.18 sec
```

File: cmake/modules/AddCephTest.cmake (hunk 3 above)
Comment (tchaikov) [reply cmake/modules/AddCephTest.cmake]: @rosinL i still suggest to add some comment to explain the the way how we map the resources in CMake. as `RESOURCE_GROUPS` is not not well-known CMake technique, a comment should help other developers to under our project uses it.

**Review by liu-chunmei — APPROVED** [review]

File: run-make-check.sh [hunk 9]
```diff
@@ -22,6 +22,14 @@ source src/script/run-make.sh
 
 set -e
 
+function gen_ctest_resource_file () {
```
Comment (tchaikov) [inline run-make-check.sh:25]: please remove the space after `gen_ctest_resource_file`.

File: src/test/crimson/ctest_utils.h [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+static std::vector<ctest_resource> parse_ctest_resources(const std::string& resource_spec) {
+  std::vector<std::string> resources;
+  boost::split(resources, resource_spec, boost::is_any_of(";"));
+  std::regex res_regex("id:([0-9]+),slots:([0-9]+)");
+  std::vector<ctest_resource> ctest_resources;
+  for (auto& resource : resources) {
+    std::smatch matched;
+    if (std::regex_match(resource, matched, res_regex) == true) {
```
Comment (tchaikov) [inline src/test/crimson/ctest_utils.h:31]:
```suggestion
    if (std::regex_match(resource, matched, res_regex)) {
```

File: cmake/modules/AddCephTest.cmake [hunk 11]
```diff
@@ -19,9 +19,27 @@ function(add_ceph_test test_name test_path)
     PATH=${CMAKE_RUNTIME_OUTPUT_DIRECTORY}:${CMAKE_SOURCE_DIR}/src:$ENV{PATH}
     PYTHONPATH=${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/cython_modules/lib.3:${CMAKE_SOURCE_DIR}/src/pybind
     CEPH_BUILD_VIRTUALENV=${CEPH_BUILD_VIRTUALENV})
-  # none of the tests should take more than 1 hour to complete
   set_property(TEST ${test_name}
     PROPERTY TIMEOUT ${CEPH_TEST_TIMEOUT})
+  # Crimson seastar unittest always run with --smp N to start N threads. By default, crimson seastar unittest
+  # will take cpu cores(from 0 to N-1), starting one thread per core. When running many crimson seastar unittests
```
Comment (tchaikov) [inline cmake/modules/AddCephTest.cmake:25]:
```suggestion
  # will take cpu cores[0, N), starting one thread per core. When running many crimson seastar unittests
```

**Review by tchaikov — APPROVED** [review]
@rosinL Rixin, we are close. lgtm modulo couple nits.

**Review by tchaikov — APPROVED** [review]
lgtm

**Comment by tchaikov** [comment]
@athanatos hey Sam, do you want to take a final look?

**Review by athanatos — APPROVED** [review]

**Comment by rosinL** [comment]
Can we merge this directly? As this pr only affect unittest and we have no qa suite to test this.

**Comment by tchaikov** [comment]
yes, agreed. merged.
